src/Groups/cosets.jl

```diff
@@ -11,20 +11,37 @@ struct GroupCoset{T<: GAPGroup, S <: GAPGroupElem}
    H::GAPGroup             # subgroup (may have a different type)
    repr::S                 # element
    side::Symbol            # says if the coset is left or right
-   X::GapObj               # GapObj(H*repr)
+   X::Ref{GapObj}          # GapObj(H*repr)
+
+   function GroupCoset(G::T, H::GAPGroup, representative::S, side::Symbol) where {T<: GAPGroup, S<:GAPGroupElem}
+     return new{T, S}(G, H, representative, side, Ref{GapObj}())
+   end
 end
 
-GAP.@install GapObj(obj::GroupCoset) = obj.X
+GAP.@install function GapObj(obj::GroupCoset)
+  if !isassigned(obj.X)
+    g = GapObj(representative(obj))
+    if is_right(obj)
+      obj.X[] = GAPWrap.RightCoset(GapObj(obj.H), g)
+    else
+      obj.X[] = GAPWrap.RightCoset(GAPWrap.ConjugateSubgroup(GapObj(obj.H), GAPWrap.Inverse(g)), g)
+    end
+  end
+  return obj.X[]
+end
 
 Base.hash(x::GroupCoset, h::UInt) = h # FIXME
 Base.eltype(::Type{GroupCoset{T,S}}) where {T,S} = S
 
-function _group_coset(G::GAPGroup, H::GAPGroup, repr::GAPGroupElem, side::Symbol, X::GapObj)
-  return GroupCoset{typeof(G), typeof(repr)}(G, H, repr, side, X)
-end
-
 function ==(x::GroupCoset, y::GroupCoset)
-   return GapObj(x) == GapObj(y) && x.side == y.side
+  (x.side == y.side && x.G == y.G && x.H == y.H ) || return false
+  if is_right(x)
+    # Hx == Hy if x/y in H
+    return representative(x) / representative(y) in x.H
+  else
+    # xH == yH if x\y in H
+    return representative(x) \ representative(y) in x.H
```

Comment:
I find the comment versus object names confusing:  $Hx$ corresponds to the Julia variable `x`; where the $H$ is `x.H` and the $x$ is `representative(x)`.

Perhaps we can name things a bit and e.g. call the Julia variables differently (say `A` and `B`), or the comment variables (e.g. `Ha` and `Hb`)

Reply:
o.k.